accepted/future-releases/variance/feature-specification.md

```diff
@@ -0,0 +1,373 @@
+`SomeClass<num>.foo(4.2)`).
+
+However, in general, every member access where a covariant type parameter
+occurs in a non-covariant position may cause a dynamic type error, because
+the actual type annotation at run time&mdash;say, the type of a parameter
+of a method&mdash;is a subtype of the one which occurs in the static type.
+
+This feature introduces explicit variance modifiers for type parameters. It
+includes compile-time restrictions on type declarations and on the use of
+objects whose static type includes these modifiers, ensuring that the
+above-mentioned dynamic type errors cannot occur.
+
+In order to ease the transition where types with explicit variance are
+created and used, this proposal allows for certain subtype relationships
+where dynamic type checks are still needed when using legacy types (where
+type parameters are _implicitly_ covariant) to access an object, even in
+the case where the object has a type with explicit variance. For example,
+it is allowed to declare `class MyList<out E> implements List<E> {...}`,
+even though this means that `MyList` has members such as `add` that require
+dynamic checks and may incur a dynamic type error.
+
+
+## Syntax
+
+The grammar is adjusted as follows:
+
+```
+<typeParameter> ::= // Modified rule.
+    <metadata> <typeParameterVariance>? <typeIdentifier>
+    ('extends' <typeNotVoid>)?
+
+<typeParameterVariance> ::= // New rule.
+    'out' | 'inout' | 'in'
+```
+
+`out` and `inout` are added to the set of built-in identifiers (* and `in`
+is already a reserved word*).
+
+
+## Static Analysis
+
+This feature allows type parameters to be declared with a _variance
+modifier_ which is one of `out`, `inout`, or `in`. This implies that the
+use of such a type parameter is restricted, in return for improved static
+type safety. Moreover, the rules for other topics like subtyping and for
+determining the variance of a subterm in a type are adjusted.
+
+
+### Subtype Rules
+
+The interface compositionality rule in [subtyping.md] is updated
+as follows:
+
+[subtyping.md]: https://github.com/dart-lang/language/blob/master/resources/type-system/subtyping.md
+
+- **Interface Compositionality**: `T0` is an interface type `C0<S0, ..., Sk>`
+  and `T1` is `C0<U0, ..., Uk>`. For `i` in `0..k`, let `vi` be the declared
+  variance of the `i`th type parameter of `C0`. Then, for each `i` in `0..k`,
+  one of the following holds:
+  - `Si <: Ui` and `vi` is absent or `out`.
+  - `Ui <: Si` and `vi` is `in`.
+  - `Si <: Ui` and `Ui <: Si`, and `vi` is `inout`.
+
+
+### Variance Rules
+
+The rules for determining the variance of a position are updated as follows:
+
+We say that a type parameter of a generic class is _covariant_ if it has no
+variance modifier or it has the modifier `out`; we say that it is
+_contravariant_ if it has the modifier `in`; and we say that it is
+_invariant_ if it has the modifier `inout`.
+
+The covariant occurrences of a type (schema) `T` in another type (schema)
```

Comment:
This definition wigs me out a bit: I know what it's saying, but if I really get pedantic it's weird.  It's basically computing a set which will have cardinality zero or one.  That is, the base case says that `S` is a covariant occurrence of `T` (that is, they are same the type variable).  Presumably we're meant to interpret this as a singleton set.  But the only thing ever put in such as set is... `T`.  So all of the unions are either of empty sets, or of the same singleton set.  So all this is computing is "yes this variable occurs" or "no it doesn't".  So I think it would be better to phrase these as predicates: "We say that a type (schema) `T` occurs covariantly in another type (schema) `S` if: ...."